source/server/listener_impl.cc

```diff
@@ -154,7 +155,8 @@ ListenerImpl::ListenerImpl(const envoy::api::v2::Listener& config, const std::st
   if (PROTOBUF_GET_WRAPPED_OR_DEFAULT(config, freebind, false)) {
     addListenSocketOptions(Network::SocketOptionFactory::buildIpFreebindOptions());
   }
-  if ((socket_type == Network::Address::SocketType::Datagram) || config.reuse_port()) {
+  if ((socket_type == Network::Address::SocketType::Datagram && concurrency > 1) ||
+      config.reuse_port()) {
```

Comment:
Won't this break restart cases in which the user still expects to be able to rebind? I think we probably need to make the option actually work for UDP also such that it can be disabled in config for the the tests?

Reply:
How does restart case work if port is 0 in config? I would suspect in the old way, we can't guarantee rebinding to the same port either. Does restarting actually depends on SO_REUSEPORT?
If that's the case, for UDP we can make it work by prioritizing reuse_port in config. If that field is set to false, RELEASE_ASSERT failure if concurrency > 1, if concurrency == 1 not set SO_REUSEPORT. In this way, a config with concurrency > 1 and reuse_port not specified or set to false will crash at loading config.

Reply:
I'm not worried about the port 0 case, just the case in which someone wants to start a new proxy with SO_REUSEPORT and then shut down the old proxy. I'm pretty sure there are people doing this, and I don't think we should break that case for UDP and concurrency 1?

Reply:
How did user achieve that? SO_REUSEPORT only allows port re-bind within same process.

Reply:
I could be wrong, but I'm pretty sure that's not true. I think it has to be part of the same process group, same owner, or something like that.

Reply:
I changed the logic to only set SO_REUSEPORT if reuse_port == true in config.